core/src/main/kotlin/ch/ergon/dope/resolvable/expression/CaseExpression.kt

```diff
@@ -0,0 +1,132 @@
+package ch.ergon.dope.resolvable.expression
+
+import ch.ergon.dope.DopeQuery
+import ch.ergon.dope.resolvable.Resolvable
+import ch.ergon.dope.validtype.BooleanType
+import ch.ergon.dope.validtype.ValidType
+
+private const val CASE = "CASE"
+private const val WHEN = "WHEN"
+private const val THEN = "THEN"
+private const val END = "END"
+private const val ELSE = "ELSE"
+
+sealed interface CaseExpression : Expression
+
+sealed interface UnaliasedCaseExpression : CaseExpression {
+    fun alias(alias: String) = AliasedCaseExpression(alias, this)
+}
+
+class AliasedCaseExpression(
+    private val alias: String,
+    private val unaliasedCaseExpression: UnaliasedCaseExpression,
+) : CaseExpression {
+    override fun toDopeQuery(): DopeQuery {
+        val caseDopeQuery = unaliasedCaseExpression.toDopeQuery()
+
+        val queryString = caseDopeQuery.queryString.trimEnd()
+            .removeSuffix(END)
+            .plus("AS `$alias` $END")
+
+        return DopeQuery(queryString, caseDopeQuery.parameters)
+    }
+}
+
+class CaseClass<T : ValidType>(val case: UnaliasedExpression<T>) : Resolvable {
+    override fun toDopeQuery(): DopeQuery {
+        val caseDopeQuery = case.toDopeQuery()
+        return DopeQuery("$CASE ${caseDopeQuery.queryString} $END", caseDopeQuery.parameters)
+    }
+}
+
+class SimpleCaseExpression<T : ValidType>(
+    val case: UnaliasedExpression<T>,
+    val firstWhenThen: Pair<TypeExpression<T>, UnaliasedExpression<out ValidType>>,
+    val whenThenMap: Map<TypeExpression<T>, UnaliasedExpression<out ValidType>> = emptyMap(),
```

Comment:
I liked the vararg approach better. Why did you decide to do Pairs and Maps?
And how do you know which type gets returned, is this type safe?

Reply:
I changed it to`vararg`. I used the Map, because I used it in a different approche where i added the firstWhenThen also to the map and used it in different constructors, but I think it makes more sense with the `vararg`.
We don't need to now which type gets returned and yes, it should be type safe.